gpt4all-backend/llamamodel.cpp

```diff
@@ -48,16 +49,7 @@ struct gpt_params {
     bool use_mlock         = false; // use mlock to keep model in memory
 };
 
-struct LLamaPrivate {
-    const std::string modelPath;
-    bool modelLoaded;
-    llama_context *ctx = nullptr;
-    llama_context_params params;
-    int64_t n_threads = 0;
-    bool empty = true;
-};
-
-static int llama_sample_top_p_top_k(
+int llama_sample_top_p_top_k(
```

Comment:
The `static` should not be removed. It has nothing to do with uniform style across implementations.

Firstly, `llama_sample_top_p_top_k()` was `static` in upstream llama.cpp before being removed in [commit dd7eff5](https://github.com/ggerganov/llama.cpp/commit/dd7eff57d8491792010b1002b8de6a4b54912e5c):

```
$ git grep -A6 'llama_vocab::id llama_sample_top_p_top_k' dd7eff5^
dd7eff5^:llama.cpp:static llama_vocab::id llama_sample_top_p_top_k(
dd7eff5^:llama.cpp-        llama_context & lctx,
dd7eff5^:llama.cpp-        const std::vector<llama_vocab::id> & last_n_tokens,
dd7eff5^:llama.cpp-        int top_k,
dd7eff5^:llama.cpp-        float top_p,
dd7eff5^:llama.cpp-        float temp,
dd7eff5^:llama.cpp-        float repeat_penalty) {
```

Secondly, when looking at the other implementations in gpt4all, i.e. [gpt4all-backend/gptj.cpp](https://github.com/nomic-ai/gpt4all/blob/914519e772fd78c15691dcd0b8bac60d6af514ec/gpt4all-backend/gptj.cpp#LL997C1-L997C1) and [gpt4all-backend/mpt.cpp](https://github.com/nomic-ai/gpt4all/blob/914519e772fd78c15691dcd0b8bac60d6af514ec/gpt4all-backend/mpt.cpp#LL922C1-L922C1), they both call a similar function named `gpt_sample_top_k_top_p()`. I assume this is what you mean when you mention aligning the style with other implementations, as `gpt_sample_top_k_top_p()` isn't `static`. There's a good non-stylistic reason for it being that way.

`gpt_sample_top_k_top_p()` is defined in [gpt4all-backend/utils.cpp](https://github.com/nomic-ai/gpt4all/blob/914519e772fd78c15691dcd0b8bac60d6af514ec/gpt4all-backend/utils.cpp#LL220C1-L220C1) and so it *has* to be made externally visible, otherwise GPTJ and MPT wouldn't be able to call it. The same doesn't apply to `llama_sample_top_p_top_k()` at all. It's not meant to be called from other translation units. It exists only for `gp4all-backend/llamamodel.cpp` internally and therefore shouldn't be externally visible. There is nothing more to it. Style and internal consistency are not relevant factors.

This became somewhat of a rant because we discussed this earlier in Discord, sorry about the wall of text. I needed to point out the stuff about GPTJ and MPT here because I never got around to that yesterday.